### ISSUE TITLE
[Awaiting Payment 11th Jan] [$500] Chat-"Hmm... it's not here" appears after clicking Previous arrow from an attachment preview

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#31105]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.97-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:** @

## Action Performed:
1. Create a request with a receipt via `Request Money > `Scan` (if you don't have one already)
2. Navigate to the IOU report 
3. Add an attachment to the IOU report via `+` > `Add attachment`
4. Click on the uploaded attachment
5. Tap the `<` arrow in the attachment carousal
6. Observe the "Hmm it's not here" page 

## Expected Result:
The uploaded receipt can be seen in the attachment carousal modal

## Actual Result:
The uploaded receipt can not be seen and "Hmm... it's not here" appears after clicking the `<` navigation arrow from an attachment preview

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/115492554/164a045a-46d9-456a-b924-8f036c45703c

![Bug6269678_1699519179590!Receipt_preview_hmm_its_not_here](https://github.com/Expensify/App/assets/115492554/988b5d5a-a0ba-40f9-ba5b-ffd4f3135f5e)

![Bug6269678_1699519463833!IMG_20231109_114404](https://github.com/Expensify/App/assets/115492554/3acd6b76-b723-482c-8b76-20fa752448ba)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a232fe720233dddb</li>
        <li>Upwork Job ID: 1722561095734079488</li>
        <li>Last Price Increase: 2023-11-09</li>
<li>Automatic offers: </li>
<ul>
<li>dukenv0307 | Contributor | 27703336</li>
<li>shubham1206agra | Contributor | 28051420</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by trjExpensify** [comment]
>We get parent report action here to compare transactionID. But the report here is IOU report that makes this check is wrong and not found page appears.

I think the proposal lacks a little specificity as to the root cause. Why does this only happen when the receipt is still scanning though? For example here I'm in a transaction thread on an `iou` report and the carousel works just fine: 

https://github.com/Expensify/App/assets/16232057/35f2fade-778d-4d30-a52d-7479c4b6ce6c

**Comment by esh-g** [comment]
# Proposal

## RCA
Parent report action transaction ID checking is done wrong.

## Changes
Change the `compareImage` function to the following:
```js
const compareImage = useCallback(
        (attachment) => {
            if (attachment.isReceipt && isReceipt) {
                const action = ReportActionsUtils.getParentReportAction(report);
                console.log(action);
                const parentTransactionIDs = Object.keys(action?.childRecentReceiptTransactionIDs)
                return _.includes(parentTransactionIDs, attachment.transactionID);
            }
            return attachment.source === source;
        },
        [source, report, isReceipt],
    );
    ```

**Comment by trjExpensify** [comment]
So @dukenv0307 why does this work fine in the transaction thread but not the IOU report?

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Chat-"Hmm... it's not here" appears after clicking Previous arrow from an attachment preview

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/components/Attachments/AttachmentCarousel/index.js#L55-L63

The logic of `compareImage` function is incorrect, It cause in line 63 we setPage(-1) --> Not Found Page.

Let see the logic of `compareImage` function

https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/components/Attachments/AttachmentCarousel/index.js#L42-L49

The logic to get the current transactionID is incorrect. It is only correct in the Transaction Thread 
<img width="1080" alt="Screenshot 2023-11-10 at 14 49 14" src="https://github.com/Expensify/App/assets/141406735/cb7fb80f-f350-4ec7-8950-ef9c21d6b3bf">

With IOU/Expense Report

<img width="1067" alt="Screenshot 2023-11-10 at 14 50 04" src="https://github.com/Expensify/App/assets/141406735/8222894b-0051-4e7c-9034-bf469a0c49f1">

https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/components/Attachments/AttachmentCarousel/index.js#L44-L46
This logic is incorrect.

It explains why this bug only happens in IOU report (not in Transaction Thread)

Note that: If we get transactionID based on report and reportAction props, for each report type, we need a separate logic to get transactionID 
### What changes do you think we should make in order to solve the problem?
We should create new state called currentTransactionID
In here
https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/components/Attachments/AttachmentCarousel/index.js#L95
we will call `setCurrentTransactionID(entry.item.transactionID)` and use `currentTransactionID` instead of this logic

https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/components/Attachments/AttachmentCarousel/index.js#L44-L46

### Result

https://github.com/Expensify/App/assets/141406735/8c627d8c-0cf2-4065-a854-ed79d0fc382e

**Comment by mananjadhav** [comment]
Checked the proposals, and  @dukenv0307's [proposal](https://github.com/Expensify/App/issues/31105#issuecomment-1803588737) looks good.

🎀 👀 🎀 C+ reviewed.

**Comment by DylanDylann** [comment]
@dukenv0307 In the yout main solution
> We should check here if the report is IOU or Expense report we shouldn't add this action to attachment list

You mean that with the IOU report, in AttachmentCarousel we will not scroll to the receipt image, we only scroll attachment

**Comment by DylanDylann** [comment]
In case, user send an attachment to IOU report. What happen?

**Comment by DylanDylann** [comment]
@dukenv0307 Thanks for your explanation, I think that we should display the receipt in the scroll list in the IOU report like we did in transaction thread
@mananjadhav I think we should confirm the expected before moving forward
@trjExpensify In the IOU report, when user click to attachment, the scroll list will display to allow user move to next/previous image. Should we display the receipt image of request in this list? Could you help to confirm the expected in this case?
cc @shawnborton Ping you here to get some opinion from UX perspective

**Comment by trjExpensify** [comment]
>@trjExpensify In the IOU report, when user click to attachment, the scroll list will display to allow user move to next/previous image. Should we display the receipt image of request in this list? Could you help to confirm the expected in this case?

I don't see a reason to not allow that. We do in the transaction thread for example as illustrated [here](https://github.com/Expensify/App/issues/31105#issuecomment-1803608943).

**Comment by DylanDylann** [comment]
@mananjadhav could you help to re-evaluate proposal based on the expected above

**Comment by mountiny** [comment]
Waiting for @mananjadhav to triage this now after more conversation

**Comment by trjExpensify** [comment]
>@trjExpensify The problem here is if we don't send any attachments in IOU report, we don't have a way to open the receipt image of the transaction, that creates an inconsistency because we can only open it by opening another attachment while in the transaction thread report we can open reciept image by clicking to this image.

I see. I think the expected behaviour of tapping the request preview component in the iou/expenseReport should still be to open the request thread not an attachment preview. So I think I see what you're saying...

- If the iou/expenseReport does not have an attachment, there's nothing to launch the carousal to cycle the receipts attached to the requests on the report. 
- If the iou/expenseReport has an attachment, you can preview that attachment and then the carousal will cycle attachments and receipts. 
- That's somewhat inconsistent behaviour to launch the carousal, as it's only possible when an attachment is uploaded to the iou/expenseReport. 
- So are we 1) okay with that? or 2) do we not include receipts in the carousal when accessible. If we opt for 2, should we make that consistent with the transaction thread, such that only attachments in the transaction thread are included in the carousal?

**Comment by esh-g** [comment]
I just wanted to ask if there is a case where the `attachment.source` would be `undefined`... because I have tried from both request sender and receiver side and the `attachment.source` is defined...

**Comment by mountiny** [comment]
I personally think we should not treat the receipt previews same as attachments in the chat. Its not the same thing, you should only be able to have a preview of the receipt by clicking the request preview and then clicking the attached receipt. I think this adds unnecessary complexity to the code and attachments and scanned receipts should not be treated the same

**Comment by trjExpensify** [comment]
Yeah cool, I'm fine for that if we apply that logic consistently. So in both the iou/expenseReport AND the transaction thread. The only items in the carousal are the uploaded attachments, not receipts.

**Comment by mountiny** [comment]
Correct, I think that makes the most sense in our usecase.

**Comment by esh-g** [comment]
## Proposal

[Updated the alternative solution](https://github.com/Expensify/App/issues/31105#issuecomment-1803613810) as per the new requirements mentioned [here](https://github.com/Expensify/App/issues/31105#issuecomment-1808604177)

**Comment by DylanDylann** [comment]
So It seems @dukenv0307's proposal match with expected. Happy with [this comment](https://github.com/Expensify/App/issues/31105#issuecomment-1806516868)

**Comment by mananjadhav** [comment]
Sorry I didn't understand what's pending from me? We're still good with @dukenv0307's [proposal](https://github.com/Expensify/App/issues/31105#issuecomment-1806516868) right?

**Comment by situchan** [comment]
Is this still reproducible? I am not able to repro on latest main.

**Comment by mountiny** [comment]
Yes I would say so

**Comment by trjExpensify** [comment]
Yup, agreed. Let's make it consistent 👍

**Comment by shubham1206agra** [comment]
Please assign me here as C+

**Comment by trjExpensify** [comment]
👋 Can we get the checklist here, please?

**Comment by mountiny** [comment]
cc @shubham1206agra

**Comment by shubham1206agra** [comment]
Wait can you generate a checklist here?
I can't see a checklist here

**Comment by shubham1206agra** [comment]
FYI original PR was reverted
And https://github.com/Expensify/App/pull/32647 was deployed yesterday only

**Comment by mountiny** [comment]
@shubham1206agra can you just copy a checklist from some other issue?

**Comment by shubham1206agra** [comment]
I can do that
I will do that a little later

**Comment by shubham1206agra** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@shubham1206agra] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/27426
* [x]  [@shubham1206agra] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/27426/files#r1429543243
* [x]  [@shubham1206agra] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Not Required as adding new test will help us to catch this problem
* [x]  [@shubham1206agra] Determine if we should create a regression test for this bug. Yes
* [x]  [@shubham1206agra] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
1. Create a Scan Request
2. Add some attachments to the money request.
3. Use arrows to navigate through attachments, and observe that no receipt is shown there. It should only been shown separately.
* [ ]  [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by trjExpensify** [comment]
Do I need to remove the payment hold here because the original PR has been reverted due to regressions?

**Comment by situchan** [comment]
There's regression from 2nd PR: https://github.com/Expensify/App/issues/33363.
It's still within regression period

**Comment by trjExpensify** [comment]
So the second PR to fix the regression from the first PR is this one: https://github.com/Expensify/App/issues/31105#issuecomment-1858110758

That has a couple of days to still go, but @situchan you're saying there's now a new regression from that 2nd PR: https://github.com/Expensify/App/issues/31105#issuecomment-1864723900

If I have that correct, I _think_ we need to remove the payment hold to await Melv adding that again on the deploy of it?

**Comment by situchan** [comment]
yes correct. @dukenv0307 should take responsible for fixing it.
And then I don't suggest to apply regression penalty given the efforts they made to fix this issue so far (just my suggestion)

**Comment by mountiny** [comment]
Thanks

**Comment by shubham1206agra** [comment]
@trjExpensify Can you add the correct labels again for payment? It is due on 11th January.

**Comment by trjExpensify** [comment]
Sure, done!

**Comment by trjExpensify** [comment]
Okay, so confirming payments as follows: 

- $250 to @shubham1206agra for the C+ review
- $250 to @dukenv0307 for the fix 

Payment has been reduced accordingly due to the regression. Sound right?

**Comment by mananjadhav** [comment]
fyi, there were more than one regressions. I had helped with the proposal reviews and the review of the first PR that caused the first regression. Subsequently during my OOO it was assigned to @shubham1206agra.

**Comment by shubham1206agra** [comment]
> yes correct. @dukenv0307 should take responsible for fixing it.
> And then I don't suggest to apply regression penalty given the efforts they made to fix this issue so far (just my suggestion)

@trjExpensify didn't you agree to not to apply regression reduction of payment here?

**Comment by situchan** [comment]
I imagine split compensation 1k into 3 people 😄

**Comment by trjExpensify** [comment]
Ah, I missed Manan reviewing the first PR. 

>@trjExpensify didn't you agree to not to apply regression reduction of payment here?

I interpreted @situchan's suggestion as not penalising for the second regression with the follow-up PR, which would result in $0 technically, which I agreed with.

**Comment by mananjadhav** [comment]
> Ah, I missed Manan reviewing the first PR.

@trjExpensify Can you accommodate me ;)

**Comment by trjExpensify** [comment]
Yep! I've updated the payment summary [here](https://github.com/Expensify/App/issues/31105#issuecomment-1889813036) to add that. Go ahead and request. 

Drawing a line in the sand, @dukenv0307 I've paid you $250. @shubham1206agra I've sent the offer for the same. Once accepted, I'll settle and close this out.

**Comment by shubham1206agra** [comment]
@trjExpensify Accepted

**Comment by trjExpensify** [comment]
Settled up!

**Comment by JmillsExpensify** [comment]
$250 payment approved for @mananjadhav based on [this comment](https://github.com/Expensify/App/issues/31105#issuecomment-1889813036).
